Mathlib/Algebra/Homology/ShortComplex/Refinements.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Homology.ShortComplex.Exact
+
+/-!
+# Refinements
+
+In order to prove injectivity/surjectivity/exactness properties for diagrams
+in the category of abelian groups, we often need to do diagram chases.
+Some of these can be carried out in more general abelian categories:
+for example, a morphism `X ⟶ Y` in an abelian category `C` is a
+monomorphism if and only if for all `A : C`, the induced map
+`(A ⟶ X) → (A ⟶ Y)` of abelian groups is a monomorphism, i.e. injective.
+Alternatively, the yoneda presheaf functor which sends `X` to the
+presheaf of maps `A ⟶ X` for all `A : C` preserves and reflects
+monomorphisms.
+
+However, if `p : X ⟶ Y` is an epimorphism in `C` and `A : C`,
+`(A ⟶ X) → (A ⟶ Y)` may fail to be surjective (unless `p` is a split
+epimorphism).
+
+In this file, the basic result is `epi_iff_surjective_up_to_refinements`
+which states that `f : X ⟶ Y` is a morphism in an abelian category,
+then it is an epimorphism if and only if for all `y : A ⟶ Y`,
+there exists an epimorphism `π : A' ⟶ A` and `x : A' ⟶ X` such
+that `π ≫ y = x ≫ f`. In order words, if we allow a precomposition
+with an epimorphism, we may lift a morphism to `Y` to a morphism to `X`.
+Following unpublished notes by George Bergman, we shall say that the
+precomposition by an epimorphism `π ≫ y` is a refinement of `y`. Then,
+we get that an epimorphism is a morphism that is "surjective up to refinements".
+(This result is similar to the fact that a morphism of sheaves on
+a topological space or a site is epi iff sections can be lifted
+locally. Then, arguing "up to refinements" is very similar to
+arguing locally for a Grothendieck topology. It is even likely that
+it is a particular case (consider the canonical topology on
+the abelian category `C`).
```

Comment:
Do you want to turn this into a TODO? Or phrase it more directly as a question that can be investigated?

Reply:
I have turned it into a TODO, with a more precise formulation.